modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTask.java

```diff
@@ -102,37 +102,79 @@ private ReindexTask(long id, String type, String action, TaskId parentTask, Map<
         this.childTask = new BulkByScrollTask(id, type, action, getDescription(), parentTask, headers);
+        this.transientStatus = childTask.getStatus();
 
     }
 
     @Override
     public Status getStatus() {
-        return childTask.getStatus();
+        return transientStatus;
     }
 
     BulkByScrollTask getChildTask() {
         return childTask;
     }
 
     private void execute(ReindexJob reindexJob) {
-        String taskId = getPersistentTaskId();
         long allocationId = getAllocationId();
-        Consumer<BulkByScrollTask.Status> committedCallback = childTask::setCommittedStatus;
-        ReindexTaskStateUpdater taskUpdater = new ReindexTaskStateUpdater(reindexIndexClient, client.threadPool(), taskId, allocationId,
-            committedCallback);
+        boolean shouldStoreResult = reindexJob.shouldStoreResult();
+        TaskManager taskManager = getTaskManager();
+
+        ReindexTaskStateUpdater taskUpdater = new ReindexTaskStateUpdater(reindexIndexClient, client.threadPool(), getPersistentTaskId(),
+            allocationId, new ActionListener<>() {
+            @Override
+            public void onResponse(ReindexTaskStateDoc stateDoc) {
+                updatePersistentTaskState(new ReindexJobState(taskId, ReindexJobState.Status.DONE), new ActionListener<>() {
+                    @Override
+                    public void onResponse(PersistentTasksCustomMetaData.PersistentTask<?> persistentTask) {
+                        if (shouldStoreResult) {
+                            taskManager.storeResult(ReindexTask.this, stateDoc.getReindexResponse(), new ActionListener<>() {
+                                @Override
+                                public void onResponse(BulkByScrollResponse response) {
+                                    markAsCompleted();
+                                }
+
+                                @Override
+                                public void onFailure(Exception ex) {
+                                    logger.info("Failed to store task result", ex);
+                                    markAsFailed(ex);
+                                }
+                            });
+                        } else {
+                            markAsCompleted();
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception ex) {
+                        logger.info("Failed to update task in cluster state to success", ex);
+                        markEphemeralTaskFailed(shouldStoreResult, ex);
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.info("Reindex task failed", e);
+                updateClusterStateToFailed(reindexJob.shouldStoreResult(), ReindexJobState.Status.DONE, e);
+            }
+        }, () -> childTask.cancel("ASSIGNMENT_CONFLICT"));
```

Comment:
Since we are not interacting with the real task, I think we can just call `childTask.onCancelled` instead to avoid making `CancellableTask.cancel` public.

Reply:
I added a comment to make it "unpublic" in the future. But right now, many places in `AbstractAsyncBulkByScrollAction` interrogate if the actual task is cancelled which will not be triggered by `onCancelled`. Ideally, it would be nice if this work (AbstractAsyncBulkByScrollAction) can be refactored to not need the task.

Just could be cancelled directly / use the "checkpointing" work for signaling state updates instead of directly modifying the task. I'm not sure how doable this is.